src/librustdoc/clean/types.rs

```diff
@@ -641,6 +643,48 @@ impl Item {
             _ => false,
         }
     }
+
+    /// Returns a `FnHeader` if `self` is a function item, otherwise returns `None`.
+    crate fn fn_header(&self, tcx: TyCtxt<'_>) -> Option<hir::FnHeader> {
+        let header = match *self.kind {
+            ItemKind::ForeignFunctionItem(_) => {
+                let abi = tcx.fn_sig(self.def_id.as_def_id().unwrap()).abi();
+                hir::FnHeader {
+                    unsafety: if abi == Abi::RustIntrinsic {
+                        intrinsic_operation_unsafety(self.name.unwrap())
+                    } else {
+                        hir::Unsafety::Unsafe
+                    },
+                    abi,
+                    constness: hir::Constness::NotConst,
+                    asyncness: hir::IsAsync::NotAsync,
+                }
+            }
+            ItemKind::FunctionItem(_) | ItemKind::MethodItem(_, _) => {
+                let def_id = self.def_id.as_def_id().unwrap();
+                let sig = tcx.fn_sig(def_id);
+                let constness =
+                    if tcx.is_const_fn(def_id) && is_unstable_const_fn(tcx, def_id).is_none() {
+                        hir::Constness::Const
+                    } else {
+                        hir::Constness::NotConst
+                    };
+                let asyncness = tcx.asyncness(def_id);
+                hir::FnHeader { unsafety: sig.unsafety(), abi: sig.abi(), constness, asyncness }
+            }
+            ItemKind::TyMethodItem(_) => {
+                let sig = tcx.fn_sig(self.def_id.as_def_id().unwrap());
+                hir::FnHeader {
+                    unsafety: sig.unsafety(),
+                    abi: sig.abi(),
+                    constness: hir::Constness::NotConst,
+                    asyncness: hir::IsAsync::NotAsync,
+                }
+            }
```

Comment:
Can you delete this branch and add `TyMethodItem` as a pattern in the previous branch? I would think `is_const_fn` would return false for `TyMethodItem`s anyway. Also, this would future-proof rustdoc for const ty methods.

Reply:
So I kinda merged both by creating a function retrieving all information except for the asyncness because `tcx.asyncness` panics when called on a `TyMethodItem`.